### PR TITLE
GODRIVER-904 Improve documentation for registering custom codecs

### PR DESCRIPTION
**Pull request by divjotarora**
I'm unsure if this is enough documentation to cover all of the essentials, but I'm putting it up for review so we can all talk about it and figure out what the holes are.

### CODE REVIEW

File: bson/bsoncodec/doc.go [hunk 1]
```diff
@@ -27,28 +27,47 @@
...
+// registered reflect.Kind as long as the value's type doesn't match a registered type or hook encoder/decoder first.
+// These methods should be used to change the behavior for all values for a specific kind.
+//
+// Registry Lookup Procedure
+//
+// When looking up an encoder in a Registry, the precedence rules are as follows:
+//
+// 1. A type encoder registered for the exact type of the value.
+//
+// 2. A hook encoder registered for an interface that is implemented by the value or by a pointer to the value.
```
Comment (iwysiu) [inline bson/bsoncodec/doc.go]: Should mention what happens if it implements multiple interfaces

Comment (divjotarora) [reply bson/bsoncodec/doc.go]: done.

Comment (divjotarora) [reply bson/bsoncodec/doc.go]: done.

File: bson/bsoncodec/doc.go [hunk 2]
```diff
@@ -27,28 +27,51 @@
...
+//
+// Registry Lookup Procedure
+//
+// When looking up an encoder in a Registry, the precedence rules are as follows:
+//
+// 1. A type encoder registered for the exact type of the value.
+//
+// 2. A hook encoder registered for an interface that is implemented by the value or by a pointer to the value. If the
+// value matches multiple hooks (e.g. the type implements bsoncodec.Marshaler and bsoncodec.ValueMarshaler), the first
+// one registered will be selected. Note that registries constructed using bson.NewRegistryBuilder driver-defined
```
Comment (iwysiu) [inline bson/bsoncodec/doc.go]: bson.NewRegistryBuilder have driver-defined hooks

Comment (divjotarora) [reply bson/bsoncodec/doc.go]: done.

**Comment by divjotarora** [comment]
Discussed with @iwysiu and @craiggwilson and decided that this is ready to go.

**Pull request merged**